Remove naked return values from `ReadBranchConfig` and `prSelectorForCurrentBranch`

jtmcg opened the pull request:
## Motivation

After some discussion with @williammartin, we agreed that the naked return values pattern is something that we'd like to avoid in our codebase and that we should make an effort to remove them when we encounter them. This is the PR to remove the named return values from `ReadBranchConfig` that I encountered while working through #10188.

I also refactored `ReadBranchConfig` for two reasons: to surface the previously hidden errors it could produce and to allow for better test coverage of its core functionality. I did the latter by separating out the `git` call and the `BranchConfig` struct creation into their own functions to leverage dependency injection of the `git` call's return value into the `BranchConfig` creation.

While making this refactor work with the rest of the tool, I stumbled upon more named return values in `prSelectorForCurrentBranch` and decided to remove these as well. This resulted in a slight reorganization of the function in a way that I believe makes it easier to follow.

## Changes

- **Remove named returns from ReadBranchConfig and surface errors**
- **Refactor ReadBranchConfig for test coverage of newly returned erros**
- **Remove named return values from prSelectorForCurrentBranch**

## Testing

I haven't touched any of the original tests besides accommodating the new `ReadBranchConfig` api, and they are all passing, so I'm decently confident that this won't introduce any breaking changes. However, to exercise all the changes, follow these steps:

1. Pull down the branch `gh pr checkout 10197`
2. Build the branch
3. Execute the `gh pr status` command

jtmcg committed:
Remove named returns from ReadBranchConfig and surface errors

jtmcg committed:
Refactor ReadBranchConfig for test coverage of newly returned erros

jtmcg committed:
Remove named return values from prSelectorForCurrentBranch

BagToad commented:
I'd be interested in hearing any key points from that discussion for my own understanding :)

Is this a design choice we're making universally or is there some qualifier that would make named return values appropriate for some situations in the CLI?

jtmcg committed:
Refactor Test_readGitBranchConfig for easier parsing

jtmcg commented:
> I'd be interested in hearing any key points from that discussion for my own understanding :)

The conversation was [here in slack](https://github.slack.com/archives/C07B978DSCX/p1735941695418149). The **tl;dr** of it is that named return values and the corresponding "naked" returns obscure information and can make code hard to grok. I explain what was confusing about `ReadBranchConfig` in the thread, but I'll add that this was silently handling errors, which wasn't super clear. To demonstrate what I mean:

```golang
// This function returns the error
func errorReturningFunction() (result Result, err error) {
    _, err = someOtherFunc()
    if err != nil {
        return // equivalent to `return Result{}, err` with unnamed returns
    } 
}

// This function doesn't return an error
func silentErrorFunction() (result Result) {
    _, err = someOtherFunc()
    if err != nil {
        return // equivalent to `return Result{}` with unnamed returns
    } 
}
```

These two functions look nearly identical except for the return signature in the function declaration. Now imagine this return is 100 lines down 😅

> Is this a design choice we're making universally or is there some qualifier that would make named return values appropriate for some situations in the CLI?

I'd advocate for yes, this should be universal. The only time I could think of named returns being ok is for really short functions, and even then there's no real benefit. If that function were to grow, we'd be in a bad state

williammartin commented:
> I'd advocate for yes, this should be universal. The only time I could think of named returns being ok is for really short functions, and even then there's no real benefit. If that function were to grow, we'd be in a bad state

Firstly, let's make a distinction between named returns and naked returns. The latter depends on the former, but you can also use named returns specifically. Named returns result in the initialisation of the variable upon function entry with the zero value.

```go
func sum(nums []int) (result int) {
    for _, n := range nums {
      result = result + n
     }
    return result // <--- still is explicit about the variable being returned
```

I don't believe there are any cases in which naked returns are the right option and I would advocate that they should all be removed.

Named returns however, do facilitate some patterns that can't be achieved otherwise, such as error handling in a defer block:

```go
func fallible() error {}
func cleanup() error {}

func doThing() err {
    defer func() {
        cleanup() <--- uh oh, cleanup can also fail but we lose the error
    }()
    
   return fallible()
}
```
  
In this example the obvious thing to do is reorganise things to cleanup only happens on error, but hopefully it doesn't take too much imagination to think of a scenario where many operations can fail and if any of them fail we need to cleanup (or perhaps we need to `recover` a panic.

Thus:

```go
func fallible() error {}
func cleanup() error {}

func doThing() (err error) {
    defer func() {
        if err != nil {           
            err = errors.Join(err, cleanup()) // <--- now we join (or ignore if `nil`) the cleanup error to have that context
        }
    }()
    
   err = fallible()
   return err
}
```
  
Hope that makes sense.

In my opinion:
 * No naked returns
 * Named returns only in pursuit of some specific patterns. There better be a clear reason for it.

williammartin commented on git/client.go:
```diff
@@ -377,19 +377,39 @@ func (c *Client) lookupCommit(ctx context.Context, sha, format string) ([]byte,
+// Downstream consumers of ReadBranchConfig should consider the behavior they desire if this errors,
+// as an empty config is not necessarily breaking.
+func (c *Client) ReadBranchConfig(ctx context.Context, branch string) (BranchConfig, error) {
+	var cfg BranchConfig
```
**Change Requested**

There's no reason to initialise this here. We only use it in the error case in which case it should be:

```go
if err != nil {
   return BranchConfig{}, err
}
```

The problem with defining things in outer scopes like this is that it implies the value is going to be mutated somewhere.

jtmcg replied:
💯 I'm with you. It's also not _that_ much different from just declaring it as a named return value.

williammartin commented on git/client.go:
```diff
@@ -377,19 +377,39 @@ func (c *Client) lookupCommit(ctx context.Context, sha, format string) ([]byte,
+func (c *Client) ReadBranchConfig(ctx context.Context, branch string) (BranchConfig, error) {
+	var cfg BranchConfig
+
+	branchConfigOutput, err := c.readGitBranchConfig(ctx, branch)
```
**Food for Thought**

Personally I'm not a fan of this private method refactor. I don't think that it improves readability by hiding the implementation and I think in the case of `createBranchConfig` I think it's actually worse coming straight after `readBranchConfig` because the name makes it seem like a mutation operation on the same git resource, when it's really just parsing a slice of strings into another type.

More nuanced, I don't really like this because I don't think the functions form a meaningful abstraction. `readGitBranchConfig` is now an available method to the rest of the client, returning a slice of strings. The only reason it returns a slice of strings is so it can pass it to another function that accepts a slice of strings.

Furthermore, `createBranchConfig` being a method on `Client` makes it appear like it needs something from the client, which it doesn't. If you want to keep the private functions, I think this would be better as a bare function called `parseBranchConfig(configLines []string)`.

Most of this is vibes about what we value, so this is non blocking.

jtmcg replied:
🤔 Maybe I went too far with this, as I appreciate your points. I think there might be somewhere in the middle that satisfies both of us.

My motivation for splitting it this way was to separate the git command from the string parsing. I can see why extracting the git command is unnecessary, as it isn't used anywhere else. I'm comfortable putting it back into the `ReadBranchConfig` method. 

As for `createBranchConfig`, I was not very intentional about whether or not these were methods on the `Client` or not, and can totally see why adding adding `createBranchConfig` as a client method is misleading... I think leaving it as a standalone helper is valuable, though (I think you're advocating for with `parseBranchConfig`). I really dislike esoteric invocations/transformations like this being buried in functions because they are hard to reason through and super specific to their inputs. Separating them out allows for a very simple to understand code block and an easy to grok utility function - especially since the tests of that function do a great job demonstrating what it's doing: give me this string and I give you this struct.

Let me back out of this a bit and see if we land somewhere that satisfies both of us.

jtmcg replied:
See this [commit](https://github.com/cli/cli/pull/10197/commits/4575692ebfbde0bfb9827e0fffea7d0751d9744f)

williammartin replied:
I like this a lot more. In fact it makes me think that `parseBranchConfig` should become exported `git.ParseBranchConfig` and live next to the struct. I could be persuaded either way though, there's pros and cons. Happy to leave it as is.

jtmcg replied:
Great! What makes you think it should be exported? I don't have much intuition about what should and shouldn't be exported and most of my thoughts around it are centered on "is this needed outside this package". Thus, I would choose to leave it unexported, here, as I'm unaware of any use-cases outside of this one.

That said, I suppose exporting it makes it more discoverable outside of the package and might reduce the chance that someone duplicates this functionality?

What do you think about when considering whether or not to export a function?

williammartin replied:
Mainly I think about module cohesiveness. How would I want to construct a `BranchConfig`? Probably I want to either construct it directly, or I want to be able to parse the lines that `git config` would have returned. Having the typical construction functions live next to the type aids in discoverability and understanding of the intention behind the type. Plus it's based on _vibes_.

I definitely don't think it's wrong to keep it unexported and promote it later. It depends how confident we are in the parsing interface being stable.

Like I said, can be persuaded either way. Do whatever you feel most confident about.

The most important thing IMO is that it's standalone, and that the other private method got re-inlined. I'm very happy to keep it as-is.

williammartin commented on git/client_test.go:
```diff
@@ -747,13 +747,121 @@ func TestClientReadBranchConfig(t *testing.T) {
 	}
 }
 
+func Test_readGitBranchConfig(t *testing.T) {
```
**Food for Thought**

I think we need to get together and hash out the testing of private functions. I'll admit I have a pretty knee jerk response to these, but you are clearly deriving some value out of them, and I need to understand what that is so that I don't have the same response on every PR.

williammartin replied:
For example, you said in the PR description that you extracted the private functions

> to allow for better test coverage of its core functionality.

Which to me screams missing confidence in the tests of the public interface, hindering refactoring.

jtmcg replied:
Let's chat about this on Friday. I think my [above comment](https://github.com/cli/cli/pull/10197#discussion_r1907574035) alludes to the value I'm deriving a bit, but it's definitely a good conversation to have.

williammartin commented on pkg/cmd/pr/create/create.go:
```diff
@@ -680,7 +680,11 @@ func NewCreateContext(opts *CreateOptions) (*CreateContext, error) {
+	headBranchConfig, err := gitClient.ReadBranchConfig(context.Background(), headBranch)
+	if err != nil {
+		// We can still proceed without the branch config, so print to stderr but continue
+		fmt.Fprintf(opts.IO.ErrOut, "Error reading git branch config: %v\n", err)
```
**Question**

Is it really an error if the branch config is empty? Seems like a valid and expected case for the most part.

jtmcg replied:
🤔 This happens with a git command failure, i.e. exit status `1`. That can happen if the command doesn't work, so an actual error interacting with `git config`, or if there's an empty return. Perhaps this isn't an "Error" and more like "Unable to read git config". 

I thought surfacing this might be helpful for those folks who are running into issues while trying to use the custom configuration tooling [we're implementing](https://github.com/cli/cli/issues/8979#issuecomment-2393677927).

jtmcg replied:
Hmm... I think I've changed my mind. I want to put this message back into the `ReadBranchConfig` method for two reasons:

1. I'm obscuring which code path it came from if one of the two possible errors are non-nil
2. I'm erroring for an empty config, even if nothing went wrong

That said, there's no IO client in the git client so I'm not sure how we want to surface this info... It seems wrong to give the git client the io client for some reason... I think this is the primary driver I had for returning the error.

I would love to hear your thoughts.

williammartin replied:
I think we want to make a distinction between:
 * There was an error working with git
 * There was no config for the branch

The first one seems like it's an error that should actually bubble up and exit the application.

The second one seems like it's a valid and supported case (hence, using `tryDetermineTrackingBranch`).

jtmcg replied:
Ah, I understand, and I think that gives me the insights to handle this...

We should return `BranchConfig{}, nil` simply if we didn't find any config for this search even though the command is exiting `0`. Otherwise, we should return `BranchConfig{}, err` if the command didn't actually run. That's going to require some groking of Stderr, but I think the info is there for me to assemble this 👍

williammartin replied:
I think so but I haven't taken a close look at how the error handling currently happens for the git commands.

jtmcg replied:
Initial testing is promising. Here's the understanding I'm operating under:

The [GitClient](https://github.com/cli/cli/blob/713346c7369f2e29963d04fd7c71cc1236a56bd9/git/client.go#L52-L62) is used to call `git` commands and a new one is instantiated for each git command we need to run*. It contains a [command context](https://pkg.go.dev/os/exec#Cmd), which holds `Stdout`, `Stderr`, `ExitCode`, and `Err`, among other things. When we execute a git command via the GitClient, the responses are stored in these various places.

Since `git config --get-regexp <regex>` will return exit code `1` but empty `Stderr` when it successfully runs but finds no matches. This results in an [error being returned by cmd.Run()](https://pkg.go.dev/os/exec#Cmd.Run).

Now [cmd.Output() handles these errors by converting them to GitErrors](https://github.com/cli/cli/blob/713346c7369f2e29963d04fd7c71cc1236a56bd9/git/command.go#L44-L51). This automatically populates `cmd.Stderr` with the error message from an `ExitError`, which is [automatically populated when Stderr is nil](https://pkg.go.dev/os/exec#Cmd.Output).

What I think this means is that if the command runs but it found no matches, it will error, but not of type `ExitError`, and so Stderr will always be `""`. Otherwise, if it exits for other reasons, it will almost always be an `ExitError` and `Stderr` will have output. I could check that our error type is `ExitError` here, but I think that's the purpose of the [error handling in cmd.Output().](https://github.com/cli/cli/blob/713346c7369f2e29963d04fd7c71cc1236a56bd9/git/command.go#L44-L51)

All this means that I'm feeling confident that my new change will handle this how we want it to.

*I haven't verified that its 1:1, but I think it's implied given the struct. Wanted to acknowledge that this is an assumption.

jtmcg replied:
I think this is resolved by https://github.com/cli/cli/pull/10197/commits/322a43feff32c0071705c8b3fd75091ffb8ebc2f. In particular, I feel confident that the error returned by `ReadBranchConfig` is something we should be surfacing, now.

williammartin commented on pkg/cmd/pr/status/status.go:
```diff
@@ -184,17 +184,17 @@ func statusRun(opts *StatusOptions) error {
-func prSelectorForCurrentBranch(gitClient *git.Client, baseRepo ghrepo.Interface, prHeadRef string, rem ghContext.Remotes) (prNumber int, selector string, err error) {
-	selector = prHeadRef
-	branchConfig := gitClient.ReadBranchConfig(context.Background(), prHeadRef)
+func prSelectorForCurrentBranch(gitClient *git.Client, baseRepo ghrepo.Interface, prHeadRef string, rem ghContext.Remotes) (int, string, error) {
```
**Food For Thought**

I'm not super confident in these changes because there is only one test.

jtmcg replied:
Fair

jtmcg replied:
I added a bunch of tests for the `prSelectorForCurrentBranch` function. Thanks for the callout, here... I definitely didn't understand it as well as I do now while initially doing this update. I actually discovered that `ReadBranchConfig` was missing a test case, too, while updating these 🙌

williammartin commented on pkg/cmd/pr/status/status.go:
```diff
@@ -184,17 +184,17 @@ func statusRun(opts *StatusOptions) error {
 	if m := prHeadRE.FindStringSubmatch(branchConfig.MergeRef); m != nil {
-		prNumber, _ = strconv.Atoi(m[1])
-		return
+		prNumber, err := strconv.Atoi(m[1])
```
**Change Requested**

I think I'd prefer to see:

```go
if err != nil {
  return 0, "", err
}
return prNumber, prHeadRef, nil
```

The problem with having a single return is that you then need to understand "in the case where this is an error, is the caller supposed to care about the values of the others?

jtmcg replied:
Yup, that makes sense and is definitely along the lines of improving the readability through removing naked returns. I appreciate how intentional that is and what it communicates

williammartin commented on pkg/cmd/pr/status/status.go:
```diff
@@ -184,17 +184,17 @@ func statusRun(opts *StatusOptions) error {
-func prSelectorForCurrentBranch(gitClient *git.Client, baseRepo ghrepo.Interface, prHeadRef string, rem ghContext.Remotes) (prNumber int, selector string, err error) {
-	selector = prHeadRef
-	branchConfig := gitClient.ReadBranchConfig(context.Background(), prHeadRef)
+func prSelectorForCurrentBranch(gitClient *git.Client, baseRepo ghrepo.Interface, prHeadRef string, rem ghContext.Remotes) (int, string, error) {
```
**Suggestion**

Since we're in here, I think it would make a lot more sense to **give** this function the `BranchConfig` rather than the git client. That is, after all, what it needs, and it would make testing much more obvious not having to stub out git commands.

jtmcg replied:
Done! Good suggestion

williammartin commented on pkg/cmd/pr/status/status.go:
```diff
@@ -184,17 +184,17 @@ func statusRun(opts *StatusOptions) error {
+		return prNumber, prHeadRef, err
 	}
 
+	var err error
```
**Food for Thought**

I believe this is a straight refactor, so I'm happy to accept it.

However, I stared at this function for a long time and I'm pretty sure that the error handling is either incorrect due to shadowing (see below) or the function signature shouldn't have originally had an error in the return. In either case we can probably rewrite this better once we understand the desired behaviour.

```go
var err error
if err := do(); err != nil { // <--- this shadows the err var
    fmt.Println(err) // so it has the value in here
}
    fmt.Println(err) // <--- but not here
```

https://go.dev/play/p/OGTGE3KL8Y8

jtmcg replied:
_Iiiinteresting_... I'll dig into this a bit more

jtmcg replied:
You were right! These errors were being shadowed. I've refactored the code so they are surfaced, now, and added tests to cover them.

williammartin replied:
Right...

The thing is that I'm not actually sure what the **expected** behaviour is. For example, if `RemoteURL` and `RemoteName` are not set, then `branchConfig` is never set. If `branchConfig` is never set, then it seems acceptable for `selector` to be the value of `prHeadRef`.

The change you made changes the behaviour because now we return an `err` in this case, when maybe a failure to find a remote for `RemoteURL` or `RemoteName` meant falling back to `selector = prHeadRef` was correct.

Does that make sense?

Maybe easier to discuss this on a call?

jtmcg replied:
Yeah, but I might be misunderstanding because I don't think I've changed the behavior... `ReadBranchConfig` may be returning an error in this case (right now, but that will change in the next commit, re: this [comment](https://github.com/cli/cli/pull/10197#discussion_r1909148397)), but when `RemoteURL` and `RemoteName` aren't set we are still returning the `prHeadRef` as the `selector`. I've added a test to demonstrate this (by passing in an empty BranchConfig{}) and twiddled with the returns a bit to make it more obvious.

jtmcg replied:
commit: https://github.com/cli/cli/pull/10197/commits/d4f7576e8baaccfdd3f7e9c4418151294d67733c

williammartin replied:
Just to be clear I'm talking about these modified error checks:

https://github.com/cli/cli/blob/d4f7576e8baaccfdd3f7e9c4418151294d67733c/pkg/cmd/pr/status/status.go#L203-L213

Previously the `err` returned would have been `nil`, meaning the calling code would have had `err == nil`:

https://github.com/cli/cli/blob/4c8d124d370f9f58d9a8d8d5a5b590f409542bbe/pkg/cmd/pr/status/status.go#L96-L99

Now you are returning the error, so the calling code will exit, instead of using the returned `prHeadRef / selector`.

jtmcg replied:
Ah, yup, that's what I was missing... 🤔

jtmcg replied:
**Decision** we're going to return `nil` for the error, here, as to not disrupt the existing behavior and add a comment explaining why.

williammartin reviewed (changes requested):
Thanks for cleaning up as you go!

jtmcg committed:
Remove private readBranchConfig method and remove parseBranchConfig from Client

I think I went too far with my previous refactor and am backing out of it.
Adding a private readBranchConfig method on the client wasn't providing
any real additional value, so I've put it back into ReadBranchConfig.

However, I think there is still value in having parseBranchConfig
(formerly createBranchConfig) as a separate util function, as it both
improves readability of ReadBranchConfig and makes parsing its purpose
easier through the bespoke tests for it.

jtmcg committed:
Change ReadBranchConfig error message in gh pr create

jtmcg committed:
Add missing test for RemoteURL in parseBranchConfig

jtmcg committed:
Refactor prSelectorForCurrentBranch and tests

Replace the git config argument in prSelectorForCurrentBranch with
the branchConfig it was used to fetch. The tests needed to be refactored
accordingly to support this change to the prSelectorForCurrentBranch API.
In addition, I've moved the test to a table test format so I can expand
the test coverage in the next commit.

jtmcg committed:
Added tests to cover prSelectorForCurrentBranch for confidence in refactor

jtmcg committed:
Remove shadowed errors from prSelectorForCurrentBranch and cover with tests

jtmcg commented:
Great review, @williammartin! I appreciate the balance of revealing nuance in your comments but asking questions that led me to figuring it out... Learned a lot, here 🙌 

I think the PR is probably close, now, but I'm looking for your thoughts on the error return for `ReadBranchConfig`, re: [this comment](https://github.com/cli/cli/pull/10197#discussion_r1907887520).

jtmcg committed:
Add test for empty BranchConfig in prSelectorForCurrentBranch

jtmcg committed:
Refine error handling of ReadBranchConfig

cmd.Output() will return an error when the git command ran successfully
but had no output. To handle this, we can check Stderr, as we expect it to
be populated for any ExitErrors or otherwise when there is a command
failure.

This allows for propagation of this error handling up the call chain, so
we are now returning errors if the call to git fails instead of just
handing off an empty BranchConfig and suppressing the errors.

Additionally, I've removed some more naked returns that I found in
pkg/cmd/pr/create.go createRun

andyfeller commented on git/client.go:
```diff
@@ -377,19 +377,36 @@ func (c *Client) lookupCommit(ctx context.Context, sha, format string) ([]byte,
 	out, err := cmd.Output()
 	if err != nil {
-		return
+		// This will error if no matches are found but the git command still ran successfully. We only
+		// want to return an error if the command failed to run, usually and ExitError, which will be
+		// indicated by output on Stderr.
+		if err.(*GitError).Stderr != "" {
+			return BranchConfig{}, err
+		}
```
**Is there any functional difference with this form versus `errors.As(...)`?**

I can't say I've seen this used in the code base before and isn't clear at first glance whether this handles type appropriately. 🤔

andyfeller replied:
I hesitate with the code as-is as its unclear if other types of errors can be raised that aren't `GitError`.

jtmcg replied:
That's... an excellent question... I toyed with `errors.As(err, ExitError)`, but it seemed weirdly repetitive given it's already called in `cmd.Output()`. I hadn't considered checking `errors.As(err, GitError)`, which is sort of what this is doing. Upon further consideration, I realize there's an unhandled `panic` possible here, too.

Thus, changing to the `errors.As(err, GitError)` pattern definitely seems better. Good callout!

jtmcg replied:
See commit https://github.com/cli/cli/pull/10197/commits/48d45d0b9db6e5be0ce8f7fe50a4cac9b01e04cc

jtmcg replied:
@williammartin pointed out the [exit codes provided by git](https://git-scm.com/docs/git-config#_description), so I think we've got our most confident fix in https://github.com/cli/cli/pull/10197/commits/322a43feff32c0071705c8b3fd75091ffb8ebc2f 🙌

andyfeller commented on git/client_test.go:
```diff
@@ -747,9 +769,87 @@ func TestClientReadBranchConfig(t *testing.T) {
+			wantCmdArgs := fmt.Sprintf("path/to/git config --get-regexp ^branch\\.%s\\.(remote|merge|gh-merge-base)$", tt.branch)
+			assert.Equal(t, wantCmdArgs, strings.Join(cmd.Args[3:], " "))
 			assert.Equal(t, tt.wantBranchConfig, branchConfig)
+			if err != nil {
+				if tt.wantError == nil {
+					t.Fatalf("expected no error but got %v", err)
+				}
+				assert.Equal(t, tt.wantError.ExitCode, err.(*GitError).ExitCode)
+				assert.Equal(t, tt.wantError.Stderr, err.(*GitError).Stderr)
+			}
```
I'm having a little mental friction reconciling this versus the majority of tests where [`wantErr`](https://github.com/search?q=repo%3Acli%2Fcli+%22wantErr%22&type=code&p=1) is typically used, which largely is a boolean or string.

I get we're wanting to ensure a specific error type and state is returned, that's not my problem.  It's more of the form this takes here as I'd expect something maybe like:

```
			if tt.wantError == nil {
				require.NoError(t, err)
			} else {
				require.Equal(t, err, tt.wantError)
			}
```

Aside from my concern about other types of errors that could be returned from this code over time, it feels like there is a simpler form of the testing assertions. 🤷

jtmcg replied:
This is about accessing the GitError as I couldn't get a test failure for the "git config runs successfully but returns no output" test case without it. That said, this pattern may have been a smell raised that the `err.(*GitError)` pattern I was using above that you called out... I'll try to refactor this with the new pattern and see what happens.

jtmcg replied:
See commit https://github.com/cli/cli/pull/10197/commits/48d45d0b9db6e5be0ce8f7fe50a4cac9b01e04cc

andyfeller commented on git/client_test.go:
```diff
@@ -747,9 +769,87 @@ func TestClientReadBranchConfig(t *testing.T) {
+				MergeBase: "gh-merge-base",
+			},
+		},
+		{
+			name: "remote, merge ref, and merge base all specified",
+			configLines: []string{
+				"branch.trunk.remote origin",
+				"branch.trunk.merge refs/heads/trunk",
+				"branch.trunk.gh-merge-base gh-merge-base",
+			},
+			wantBranchConfig: BranchConfig{
+				RemoteName: "origin",
+				MergeRef:   "refs/heads/trunk",
+				MergeBase:  "gh-merge-base",
+			},
+		},
```
I like the way this is ordered relative to the above tests; if either of those fail, then there would be errors before this case.

andyfeller commented on pkg/cmd/pr/create/create.go:
```diff
@@ -266,14 +266,14 @@ func NewCmdCreate(f *cmdutil.Factory, runF func(*CreateOptions) error) *cobra.Co
 func createRun(opts *CreateOptions) (err error) {
```
Should the named return be removed, too?

```suggestion
func createRun(opts *CreateOptions) error {
```

jtmcg replied:
Yeah, that change got stranded on my local machine. I've put it into the correct commit, now. Thanks!

andyfeller commented on pkg/cmd/pr/create/create_test.go:
```diff
@@ -1717,7 +1717,7 @@ func Test_tryDetermineTrackingRef(t *testing.T) {
 				GitPath: "some/path/git",
 			}
-			headBranchConfig := gitClient.ReadBranchConfig(ctx.Background(), "feature")
+			headBranchConfig, _ := gitClient.ReadBranchConfig(ctx.Background(), "feature")
```
Why swallow the error here instead of asserting it is nil or not?

jtmcg replied:
~~Because I missed it 😅 Thanks!~~

Nope, this is correct. The branch config is set up by the `cmdStubs` in the test, so I expect this to never fail.

That said, I don't know if we actually need to call `ReadBranchConfig` in this test or if I can just use a stubbed `BranchConfig{}` instead. The latter would be much more desirable... Will investigate

jtmcg replied:
🔥 It worked (see commit https://github.com/cli/cli/pull/10197/commits/e4320ccc4bf39dd3cf0e8eba2a6872b6d42eccca)! Super happy to remove this dependency in this test, here 👍 Nice catch!

andyfeller commented on pkg/cmd/pr/shared/finder.go:
```diff
@@ -59,7 +59,8 @@ func NewFinder(factory *cmdutil.Factory) PRFinder {
 		progress:   factory.IOStreams,
 		branchConfig: func(s string) git.BranchConfig {
-			return factory.GitClient.ReadBranchConfig(context.Background(), s)
+			branchConfig, _ := factory.GitClient.ReadBranchConfig(context.Background(), s)
```
Should this panic if an error is returned?

jtmcg replied:
🤔 That's a great question... It shouldn't, and it should probably return the error now that it's good and properly handled...

jtmcg replied:
Sooooooo this looks like a can of worms 😅 I've done the absolute minimum to surface this error in the finder for now, but at first blush [parseCurrentBranch in shared/finder.go](https://github.com/cli/cli/blob/21a1f8e8a9201143b11cb8b8e5d78bc163839c60/pkg/cmd/pr/shared/finder.go#L235-L283) looks awfully similar [prSelectorForCurrentBranch pr/status/status.go](https://github.com/cli/cli/blob/21a1f8e8a9201143b11cb8b8e5d78bc163839c60/pkg/cmd/pr/status/status.go#L187-L222)...

I'm not gonna investigate that similarity further at this time

andyfeller commented on pkg/cmd/pr/status/status_test.go:
```diff
@@ -376,30 +377,209 @@ Requesting a code review from you
+		{
+			name:         "Empty branch config",
+			branchConfig: git.BranchConfig{},
+			prHeadRef:    "Frederick888/main",
```
Nothing to change but I wonder whether we want valid GitHub usernames in our tests or not.  I know plenty of other tests use placeholders rather than any meaningful user/owner account and repo.  $0.02

jtmcg replied:
This is definitely copy/pasta and an easy fix. I think it came from the contributor who originally wrote this bit of code, as it is their user name 😄 I guess that's one way to leave your mark on a codebase

andyfeller commented on pkg/cmd/pr/status/status_test.go:
```diff
@@ -376,30 +377,209 @@ Requesting a code review from you
+				MergeRef: "refs/pull/666/head",
+			},
+			prHeadRef:    "Frederick888/main",
+			wantPrNumber: 666,
```
I'm not offended by the specific content used here, but I don't think it is appropriate for an OSS project like this.

jtmcg replied:
Fair

andyfeller reviewed (commented):
Quite a bit of extensive work, but I appreciate chipping away at named returns! 🙌 

There are a few sections I had questions while a few others that raised more meta questions regarding content we use in the tests.

jtmcg committed:
Refactor error handling in ReadBranchConfig to avoid panic

jtmcg committed:
Directly stub headBranchConfig in Test_tryDetermineTrackingRef

jtmcg committed:
Surface and handle error from ReadBranchConfig in parseCurrentBranch

I've only added the one test for parseCurrentBranch because the function
appears to be largely exercised by TestFind. There's definitely an
opportunity for a bigger refactor of the tests, here, but I want to avoid
scope creep as I propagate the ReadBranchConfig api changes throughout the
codebase

jtmcg committed:
Change pr number in test

jtmcg committed:
Rename test user in tests

williammartin commented on git/client.go:
```diff
@@ -377,19 +377,38 @@ func (c *Client) lookupCommit(ctx context.Context, sha, format string) ([]byte,
+		var gitError *GitError
+		if errors.As(err, &gitError) {
+			if gitError.Stderr != "" {
+				return BranchConfig{}, err
```
I think this _probably_ works as you outlined, so I'm happy to leave it as-is. I do wonder whether we could rely on the exit code being `1`.

https://git-scm.com/docs/git-config#_description

jtmcg replied:
Ah, because the other exit codes will represent the errors that aren't from an empty return? I think that is probably better, for sure! Easier for testing as well

jtmcg replied:
Found a very similar implementation in [Remotes](https://github.com/cli/cli/blob/4c8d124d370f9f58d9a8d8d5a5b590f409542bbe/git/client.go#L182-L186) above this

jtmcg replied:
Updated in https://github.com/cli/cli/pull/10197/commits/322a43feff32c0071705c8b3fd75091ffb8ebc2f

jtmcg committed:
Change error handling on ReadBranchConfig to respect git Exit Codes

jtmcg committed:
Restore old error functionality of prSelectorForCurrentBranch

Before this refactor, the errors emitted by ghrepo.FromURL and
rem.FindName() were suppressed. It isn't clear whether this was
intentional or not, but we've made the decision here to maintain the
original error behavior while still refactoring the return values for more
clarity. I've left a comment at each error handling block to explain this
decision.

Additionally, I've added the necessary git command stubs to the other
tests in status_test.go so that the tests are now passing.

andyfeller reviewed (approved):
LGTM ✨ That said, will leave it to you, @jtmcg, if you want to wait on final review from Will 👍

williammartin reviewed (approved):
LGTM.

```
➜  cli git:(jtmcg/remove-named-returns) set -o pipefail && GH_ACCEPTANCE_HOST=github.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -coverprofile=coverage.out -coverpkg=./... -run ^TestPullRequests$ github.com/cli/cli/v2/acceptance | tparse --all go test
┌────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │                        TEST                        │             PACKAGE               │
│─────────┼─────────┼────────────────────────────────────────────────────┼───────────────────────────────────│
│  PASS   │   13.73 │ TestPullRequests/pr-create-from-issue-develop-base │ github.com/cli/cli/v2/acceptance  │
│  PASS   │   13.35 │ TestPullRequests/pr-checkout-with-url-from-fork    │ github.com/cli/cli/v2/acceptance  │
│  PASS   │   13.32 │ TestPullRequests/pr-merge-rebase-strategy          │ github.com/cli/cli/v2/acceptance  │
│  PASS   │   13.24 │ TestPullRequests/pr-merge-merge-strategy           │ github.com/cli/cli/v2/acceptance  │
│  PASS   │   11.04 │ TestPullRequests/pr-view-same-org-fork             │ github.com/cli/cli/v2/acceptance  │
│  PASS   │   10.57 │ TestPullRequests/pr-create-with-metadata           │ github.com/cli/cli/v2/acceptance  │
│  PASS   │   10.10 │ TestPullRequests/pr-create-basic                   │ github.com/cli/cli/v2/acceptance  │
│  PASS   │   10.09 │ TestPullRequests/pr-create-from-manual-merge-base  │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    9.94 │ TestPullRequests/pr-view                           │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    9.09 │ TestPullRequests/pr-list                           │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    8.43 │ TestPullRequests/pr-create-without-upstream-config │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    8.34 │ TestPullRequests/pr-checkout-by-number             │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    7.99 │ TestPullRequests/pr-comment                        │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    7.27 │ TestPullRequests/pr-checkout                       │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.00 │ TestPullRequests                                   │ github.com/cli/cli/v2/acceptance  │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │             PACKAGE              │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼──────────────────────────────────┼───────┼──────┼──────┼───────│
│  PASS   │ 25.23s  │ github.com/cli/cli/v2/acceptance │ 15.7% │  15  │  0   │  0    │
└────────────────────────────────────────────────────────────────────────────────────┘
```

williammartin commented:
I think this is probably going to cause merge conflicts with https://github.com/cli/cli/pull/9208/files#diff-ccedca01e076cd065987253b366873ca2bee2e28c9d0bc13d50d7b7795c0bf4eR380-R420

Let's fix this for @Frederick888. Getting that PR through is next on our backlog too, so maybe I'll just do it.
